Support keyed children

fitzgen opened the pull request:
@alexcrichton @tschneidereit: I'd appreciate some eyes on this PR (particularly the third commit) :eyes: 

Using keyed children gives users more control over whether particular physical nodes are reused, moved, deleted, or created afresh. Specifically, if the old virtual node and the new virtual node have the same key, then the physical DOM node must be reused, and if they don't then the old physical DOM node cannot be reused.

Presumably there is some sort of CSS transition animation or third party JS that makes the virtual DOM diffing algorithm observable, or else users would let the runtime do whatever it thought was best.

This is loosely based on Inferno's keyed patching implementation. However, we have to modify the algorithm since we are compiling the diff down into change list instructions that will be executed later, rather than applying the changes to the DOM directly as we compare virtual DOMs.

https://github.com/infernojs/inferno/blob/36fd96/packages/inferno/src/DOM/patching.ts#L530-L739

fitzgen committed:
Have a fast path for removing all children

fitzgen committed:
Diff nodes by reference rather than by value

This avoids a bunch of memcpy'ing since node isn't a tiny structure.

fitzgen committed:
Support keyed children

Using keyed children gives users more control over whether particular physical
nodes are reused, moved, deleted, or created afresh. Specifically, if the old
virtual node and the new virtual node have the same key, then the physical DOM
node must be reused, and if they don't then the old physical DOM node cannot be
reused.

Presumably there is some sort of CSS transition animation or third party JS that
makes the virtual DOM diffing algorithm observable, or else users would let the
runtime do whatever it thought was best.

This is loosely based on Inferno's keyed patching implementation. However, we
have to modify the algorithm since we are compiling the diff down into change
list instructions that will be executed later, rather than applying the changes
to the DOM directly as we compare virtual DOMs.

https://github.com/infernojs/inferno/blob/36fd96/packages/inferno/src/DOM/patching.ts#L530-L739

alexcrichton commented on src/builder.rs:
```diff
@@ -220,6 +226,56 @@ where
+    /// Set this element's key. The key must be less than or equal to `2^31 -
+    /// 1`.
+    ///
+    /// This element's parent must have `has_keyed_children(true)` set.
```
I'm still likely naive at this point (I don't currently have a grasp on what keyed children are used for) but it seems unfortunate from an API-design perspective that this property needs to be upheld. 

Would it be possible to have debug asserts for this though throughout just to make sure test builds try to catch this?

fitzgen replied:
Yeah I can def add debug asserts. It is a somewhat unfortunate constraint, but my hope is that `typed-html` or whatever would do this automatically 99% of the time.

The reason it is this way is because checking for keys is an O(n) scan, but the person rendering the virtual dom tree just *knows* whether they are using keys or not, so it seems absolutely silly to do that linear search.

But also if there are keyed children, then they should all be keyed, so probably we should do:

* fast check for if first child is keyed
* debug assert that all children are keyed or all children are not keyed
* remove the has-keyed-children flag, `bitflags` dep, and `ElementNodeFlags`
* just store a `u32` key in `ElementNode` and remove the you-can't-use-the-top-bit restriction

fitzgen replied:
Done!

alexcrichton commented on src/builder.rs:
```diff
@@ -220,6 +226,56 @@ where
+    }
+
+    /// Set this element's key. The key must be less than or equal to `2^31 -
+    /// 1`.
```
Perhaps mention that in layman's terms this means the top bit can't be used?

fitzgen replied:
Only `u32::MAX` is reserved now, and this is documented and asserted.

alexcrichton commented on src/diff.rs:
```diff
@@ -2,40 +2,47 @@ use crate::{
+//
+//     [... node]
+//
+// The stack is in the same state when this function exits.
```
Without having a deep understanding of dodrio, I'm not entirely certain what stack this is referring to. Is this the stack in the change list builder which is used as sort of a stack machine for all the opcodes generated? (I presume so, just wanted to confirm!)

fitzgen replied:
Yeah this is the interpreter's stack if it has evaluated all the emitted instructions thus far. Happy for suggestions of more precise but less wordy ways of saying this :)

fitzgen replied:
Calling this "the change list stack" instead of "the stack" to clarify here.

alexcrichton replied:
Nah I think this is fine as it was (I'll probably see an update soon anyway), mostly just wanted to double-check my own understanding!

alexcrichton commented on src/diff.rs:
```diff
@@ -2,40 +2,47 @@ use crate::{
             &NodeKind::Text(TextNode { text: new_text }),
             &NodeKind::Text(TextNode { text: old_text }),
         ) => {
-            debug!("  both are text nodes");
```
Did these end up generating too much traffic in the debug logs? (just curious as to the removal)

fitzgen replied:
Yeah. Essentially, I found that the only logs I really care about are which instructions get emitted.

alexcrichton commented on src/diff.rs:
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
-            change_list.append_child();
+        }
+        Ordering::Equal => {
+            change_list.pop();
```
I was trying to pair this up with another "push" in this function, and it took me a bit following it to figure out what was going on. Could the `i == 0` case be pulled out of the loop and listed inline above? I think the `i == 0` is guaranteed to fun due to the nonempty assertions for both arrays, but it may be a bit more clearer to pull it out to see what push this is matching.

I also don't really know what's being pushed and popped, so feel free to say "not necessary" :)

fitzgen replied:
Yeah because `new` and `old` are non-empty, we know we execute the loop at least once, and therefore push a child onto the stack, and that is what is being popped here. I can make a comment to this effect to clear things up.

alexcrichton commented on src/diff.rs:
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
+    for child in old {
+        registry.remove_subtree(child);
+    }
+    // Fast way to remove all children: set the node's textContent to an empty
```
Learn something new every day!

alexcrichton commented on src/diff.rs:
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
+    // Map from each `old` node's key to its index within `old`.
+    let mut old_key_to_old_index = FxHashMap::default();
+    old_key_to_old_index.reserve(old.len());
+    old_key_to_old_index.extend(old.iter().enumerate().map(|(i, o)| (o.key(), i)));
```
This brings up an interesting idea in the sense that it's the onus of the user to not reuse keys, right?

(should that be added to the `key` docs?

alexcrichton replied:
Actually now I'm slightly confused, it looks like `u32::MAX` is used for non-present keys, so that just means that `u32::MAX` as a key in this map is overwritten multiple times? I guess though non-max values are supposed to be unique?

fitzgen replied:
Yeah I should add an assertion for this.

alexcrichton commented on src/diff.rs:
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
+    // If none of the old keys are reused by the new children, then we
+    // remove all the remaining old children and create the new children
+    // afresh.
+    if shared_suffix_count == 0 && shared_keys.is_empty() {
```
This condition is throwing me a bit for a spin, I am naively expecting it to *just* be checking the `shared_keys` set, but I can't quite figure out why `shared_suffix_count` is factored in here?

Should `shared_suffix_count == 0` only be used to optimize which opcodes are emitted? I'd expect that we'd still have some optimized path for shared `shared_suffix_count` is nonzero but there's no shared keys

fitzgen replied:
Your intuition that there should be a fast path regardless of `shared_suffix_count`'s value is correct. The reason it doesn't exist is simply that we don't currently have instructions/functions for mass deleting and inserting in the middle of the children, only all of the children, or the latter half of the children.

It is somethign that I feel can be a follow up PR.

fitzgen replied:
Your intuition that there should be a fast path regardless of shared_suffix_count's value is correct. The reason it doesn't exist is simply that we don't currently have instructions/functions for mass deleting and inserting in the middle of the children, only all of the children, or the latter half of the children.

It is somethign that I feel can be a follow up PR.

alexcrichton commented on src/diff.rs:
```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
+    // children. Remove from the end first so that we don't mess up indices.
+    let mut removed_count = 0;
+    for (i, old_child) in old.iter().enumerate().rev() {
+        if !shared_keys.contains(&old_child.key()) {
```
Hm ok I may again be very confused, but this sort of looks like it's preserving everything with a key of `u32::MAX`?

alexcrichton reviewed (commented):
Ok so I followed everything pretty well up to the point that I hit `diff_keyed_middle`. Man that function is gnarly! I'm gonna leave the comments I have for now, but I suspect you'll have some comments which help clear up a few things for me and I can take another stab later to finish it :)

fitzgen committed:
Remove `has_keyed_children(true)` requirement for keying children

fitzgen committed:
interpreter: Clear the temporaries after applying the diff

fitzgen committed:
Assert that keys are unique among siblings in debug mode

fitzgen committed:
Clarify which stack the diffing comments are talking about

fitzgen committed:
Add comments clarifying logic about change list stack in `diff_non_keyed_children`

axelf4 commented:
Just alerting you to the fact that there exists some overlap here between this and [a diff crate](https://github.com/axelf4/lis/blob/c695f0e5b59ead9d43e0b67a172945e4d07da4f2/src/lib.rs#L244:L336) I wrote.

alexcrichton commented:
So I started off today trying to get a better idea of the middle diffing function, and ended up spending basically all day following @fitzgen down profiling rabbit holes.

At the beginning of the day though we concluded that the iteration here in the middle diffing can probably be simplified to not use chunks but rather just in-order reverse iteration. I think that's probably the next step here!

fitzgen committed:
Shrink the size of `NodeKind` by putting `ElementNode`s behind a ref

`ElementNode`s are not small, and by putting them behind a reference, then
`NodeKind`'s size drops to three words and we avoid a bunch of `memcpy`ing.

Additionally, use recursion in `EventsRegistry::remove_subtree` isntead of
maintaining an explicit work stack. Recursion is faster and we don't need to
worry about blowing the stack here since we already use recursion anyways in
diffing.

fitzgen committed:
Use recursion for `Cached<R>` tracing

It is faster than maintaining an explicit work stack.

fitzgen committed:
Update dependencies

fitzgen committed:
Simplify `diff_keyed_middle`

Remove the whole "segments" thing and just walk backwards through the `new`
children. This also lets us remove the did-we-push-a-child-or-not-yet boolean
and simplifies a bunch of stuff.
